File: 084SortedSquaredArray/sorted_squared_array.rb
# Solution 1
# O(nlog(n)) time / O(n) space

def sorted_squared_array(arr)
  sorted_squares = Array.new(arr.length, 0)

  (0..arr.length - 1).each { |idx|
    value = arr[idx]
    sorted_squares[idx] = value.to_i * value.to_i
  }
  sorted_squares.sort()
end

# Solution 2
# O(n) time / O(n) space

def sorted_squared_array(arr)
  sorted_squares = Array.new(arr.length, 0)
  smallest_value_idx = 0
  largest_value_idx = arr.length - 1

  (arr.length - 1).step(0, -1) do |idx|
    smaller_value = arr[smallest_value_idx]
    larger_value = arr[largest_value_idx]

    if smaller_value.abs > larger_value.abs
      sorted_squares[idx] = smaller_value.to_i * smaller_value.to_i
      smallest_value_idx += 1
    else
      sorted_squares[idx] = larger_value.to_i * larger_value.to_i
      largest_value_idx -= 1
    end
  end
  sorted_squares
end

File: 084SortedSquaredArray/sortedSquaredArray.py
# Solution 1
# O(nlog(n)) time / O(n) space
# n - length of the array

def sortedSquaredArray(array):
    sortedSquares = [0 for _ in array]

    for idx in range(len(array)):
        value = array[idx]
        sortedSquares[idx] = value * value

    sortedSquares.sort()
    return sortedSquares


# Solution 2
# O(n) time / O(n) space
# n - length of the array

def sortedSquaredArray(array):
    sortedSquares = [0 for _ in array]
    smallestValueIdx = 0
    largestValueIdx = len(array) - 1

    for idx in reversed(range(len(array))):
        smallerValue = array[smallestValueIdx]
        largerValue = array[largestValueIdx]

        if abs(smallerValue) > abs(largerValue):
            sortedSquares[idx] = smallerValue * smallerValue
            smallestValueIdx += 1
        else:
            sortedSquares[idx] = largerValue * largerValue
            largestValueIdx -= 1

    return sortedSquares
